### PR TITLE
Add Waf RateBasedRule support

### PR DESCRIPTION
**Pull request by bekbulatov**
Updated version of https://github.com/terraform-providers/terraform-provider-aws/pull/1285. 
I had to add the required `type` field to the rules block under `aws_waf_rule`. The value could be either `REGULAR` or `RATE_BASED` regarding which type of rule you are using.
The docs are updated as well.

### CODE REVIEW

File: aws/resource_aws_waf_rate_based_rule.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+								}
+								return
+							},
+						},
+					},
+				},
+			},
+			"rate_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (radeksimko) [inline aws/resource_aws_waf_rate_based_rule.go]: It seems that this field isn't updatable (based on API reference), shouldn't it be marked as `ForceNew` then?

File: aws/resource_aws_waf_rate_based_rule.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		return conn.DeleteRateBasedRule(req)
+	})
+	if err != nil {
+		return fmt.Errorf("Error deleting WAF Rate Based Rule: %s", err)
+	}
+
+	return nil
+}
+
+func updateWafRateBasedRuleResource(id string, oldP, newP []interface{}, rateLimit *int64, conn *waf.WAF) error {
```
Comment (radeksimko) [inline aws/resource_aws_waf_rate_based_rule.go]: It might be more convenient to accept `rateLimit int` here instead and do all the pointer magic inside the function to reduce the repetition, but no biggie.

File: aws/resource_aws_waf_web_acl.go [hunk 3]
```diff
@@ -64,6 +64,18 @@ func resourceAwsWafWebAcl() *schema.Resource {
...
+							Required: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := v.(string)
+								if value != waf.WafRuleTypeRegular && value != waf.WafRuleTypeRateBased {
+									errors = append(errors, fmt.Errorf(
+										"%q must be one of %s | %s", k, waf.WafRuleTypeRegular, waf.WafRuleTypeRateBased))
+								}
+								return
+							},
+						},
```
Comment (radeksimko) [inline aws/resource_aws_waf_web_acl.go]: Is there any particular reason why this change should be included in this PR? How is it related to WAF Rate Based Rule?

File: aws/resource_aws_waf_rate_based_rule.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"negated": &schema.Schema{
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"data_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
```
[the remark this answers is not in the document]
Comment (bekbulatov) [reply aws/resource_aws_waf_rate_based_rule.go]: Sure, I will update the schema for `Rule` as well

File: aws/resource_aws_waf_rate_based_rule.go (hunk 1 above)
Comment (bekbulatov) [reply aws/resource_aws_waf_rate_based_rule.go]: `Rate_key` could be only `IP` now. Why do you think it should be recreated if changed?

File: aws/resource_aws_waf_rate_based_rule.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+			"rate_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if value != waf.RateKeyIp {
+						errors = append(errors, fmt.Errorf("%q must be only %s", k, waf.RateKeyIp))
+					}
+					return
+				},
```
[the remark this answers is not in the document]
Comment (bekbulatov) [reply aws/resource_aws_waf_rate_based_rule.go]: Agreed, anyway the value will be validated by AWS.

File: aws/resource_aws_waf_web_acl.go (hunk 3 above)
Comment (bekbulatov) [reply aws/resource_aws_waf_web_acl.go]: I mentioned this in https://github.com/terraform-providers/terraform-provider-aws/pull/1285. By default, WebACL works with `REGULAR` rules and you will get an error if you pass `rule_id` of a `RATE_BASED` rule. This is documented in https://github.com/aws/aws-sdk-go/blob/master/service/waf/api.go#L5708-L5713

File: aws/resource_aws_waf_rate_based_rule.go (hunk 1 above)
Comment (radeksimko) [reply aws/resource_aws_waf_rate_based_rule.go]: well, you're right that there's no way to change it, but if the user is unaware they might try and at the moment we'd just ignore the field (in `Update`) and support user's assumption that the field was successfully changed.

If we instead delegate the validation work to `Create` then it will fail and user will also realize during `plan` that this can't be changed.

File: aws/resource_aws_waf_rate_based_rule.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply aws/resource_aws_waf_rate_based_rule.go]: Yep, I think the AWS API validation is the best choice here.

File: aws/resource_aws_waf_web_acl.go (hunk 3 above)
Comment (radeksimko) [reply aws/resource_aws_waf_web_acl.go]: I see, but I think we can't just introduce a `Required` field - that would break it for users with existing configs which don't have this field there.
This applies to your other change in `aws/resource_aws_waf_rule.go` too.

Comment (radeksimko) [reply aws/resource_aws_waf_web_acl.go]: Do you mind changing it to Optional?

Comment (radeksimko) [reply aws/resource_aws_waf_web_acl.go]: btw. you might be able to avoid changing the `updateWebAclResource` if you also add the current default value as `Default` - that way it should just work the same way as before, yet allow users to specify `type = "RATE_BASED"` if they need to.

Comment (bekbulatov) [reply aws/resource_aws_waf_web_acl.go]: @radeksimko Thank you for the review. I believe the `Required` change for `aws/resource_aws_waf_rule.go` is fine, as we just duplicate AWS validation, it's not a breaking change.
I made `type` optional and default to `REGULAR` as recommended.

Comment (radeksimko) [reply aws/resource_aws_waf_web_acl.go]: You're right - just checked the [API docs](http://docs.aws.amazon.com/waf/latest/APIReference/API_Predicate.html) - it was probably just overlooked.

Thanks.
